vowpalwabbit/autolink.h

```diff
@@ -1 +1,24 @@
+#pragma once
+
+#include "reductions.h"
+
 LEARNER::base_learner* autolink_setup(VW::config::options_i& options, vw& all);
+
+namespace VW
+{
+  struct autolink
```

Comment:
Why do we want this struct in the header file?  If it's for testing purposes, shouldn't there be some tests?

Reply:
It's in the header so it can be independently instantiated as part of the library. I don't really understand enough semantically what this reduction does to know what I could test here.

Reply:
I believe this is not the object we would want to instantiate.  Instead, it would be the learner which contains autolearn.

Reply:
A learner contains driver specific functionality, such as outputting and finishing examples. The reduction itself seems to be the core logic that is actually useful outside of the command line.

Reply:
I think my worry here is about type structure.  

"A learner is an object that instantiates learn() and predict()" may not be the right structure since you cannot create a collection due to nonuniform typing.

Reply:
Yeah I agree with that. There needs to be some sort of uniform type that can be used to reason here. Rajan, Peter and myself have discussed for some time the ideas of reductions being typed by what they support. (input/output, label type, pred type etc) There is definitely more work to come there, this is just a baby step to help get rid of all